makefile_include.mk

```diff
@@ -13,9 +13,13 @@ ifndef CROSS_COMPILE
   CROSS_COMPILE:=
 endif
 
-ifeq ($(CC),cc)
+ifeq ($(origin CC),default)
+ifneq (,$(shell echo $$"\#ifdef __clang__\nCLANG\n\#endif\n" | $(CC) -E - | grep CLANG))
```

Comment:
uhm, this creates the following for me:
```
$ make
<stdin>:3:2: error: #endif without #if
...
$ gcc -dumpversion
5.4.0
$ cc -dumpversion
5.4.0
$ clang -dumpversion
4.2.1
```

Ubuntu Xenial-based distro

Reply:
To add to the confusion, macOS, which uses Clang as the default compiler, seems to be fine both with and without `$$`. 🤔

Reply:
Would that be an option?
```
-ifneq (,$(shell echo $$"\#ifdef __clang__\nCLANG\n\#endif\n" | $(CC) -E - | grep CLANG))
+ifneq (,$(shell $(CC) --version | grep -i clang))
```

Reply:
The problem there is that Clang's help doesn't tell you it's Clang on every platform. For example, Clang on macOS calls itself "Apple LLVM."

I was chatting with a colleague about the `$$` conundrum and he said that it stems from the fact that `shell` in a Makefile refers to a different shell on different systems, and that it's always POSIX `sh` on FreeBSD and `dash` on Ubuntu. While an interesting tidbit, it doesn't provide an immediate solution. :wink: 

I think the easiest way forward here would be to put the check in a file, e.g. just make a file containing
```c
#ifdef __clang__
CLANG
#endif
```
and in the Makefile call `$(shell $(CC) -E file | grep CLANG)`. The downside of that of course is then there's just some annoying little file sitting floating around in the source tree...

Reply:
Figured it out. We can just conditionally add extra escaping on FreeBSD.